### PR TITLE
Frogger game

### PR DESCRIPTION
**Pull request by smillims**
### Object-Oriented JavaScript

[Demo](https://smillims.github.io/frontend-nanodegree-arcade-game/)
[Code base](https://github.com/smillims/frontend-nanodegree-arcade-game)

The code is submitted in a dedicated feature branch.
Please, review.

### CODE REVIEW

File: submissions/smillims/oop/app.js [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+// Update the enemy's position, required method for game
+// Parameter: dt, a time delta between ticks
+Enemy.prototype.update = function(dt) {
+    // You should multiply any movement by the dt parameter
+    // which will ensure the game runs at the same speed for
+	 // all computers.
+	 this.x += this.speed * dt;
+	 if (this.x > ctx.canvas.width + block.width){
+		 this.x = -101 * 2;
```
Comment (OleksiyRudenko) [inline submissions/smillims/oop/app.js:30]: `-101` resembles negated `block.width` or it is just a coincidence?

Check your entire code for mysterious numbers. They definitely have semantic meaning, thus worth defining as contants.

File: submissions/smillims/oop/app.js [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+const player = new Player(200, 400);
+
+const enemy1 = new Enemy(0, 60);
+const enemy2 = new Enemy(0, 140);
+const enemy3 = new Enemy(0, 220);
+const allEnemies = [
+	enemy1,
+	enemy2,
+	enemy3
+];
```
Comment (OleksiyRudenko) [inline submissions/smillims/oop/app.js:110]: Too much repetition. Imagine, we need creating 10 enemies.
Consider the following example:
```
const objects = [100, 150, 200].map(pos => new Circle(pos, 0));
```
Intentionally oversimplified, please do not copypaste and reconsider all variables naming.

File: submissions/smillims/oop/app.js [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+}
+
+// Now write your own player class
+// This class requires an update(), render() and
+// a handleInput() method.
+
+const Player = function(x, y, player) {
+	this.x = x;
+	this.y = y;
+	this.player = player;
```
Comment (OleksiyRudenko) [inline submissions/smillims/oop/app.js:56]: Why do you need `this.player`? It doesn't seem to be used anywhere in `Player`'s methods. Besides if this is a selfreference then `this` is already a selfreference.

File: submissions/smillims/oop/app.js [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+	 }
+};
+
+// Draw the enemy on the screen, required method for game
+Enemy.prototype.render = function() {
+    ctx.drawImage(Resources.get(this.sprite), this.x, this.y);
+};
+
+const checkCollisions = () => {
+	allEnemies.forEach(enemy => {
```
Comment (OleksiyRudenko) [inline submissions/smillims/oop/app.js:40]: Classes should not refer to any global variables, like global variable `player`, which is an instance of `Player` class (referring to global constants and globals provided by the gaming platform like `Resources` is OK).

**Review by OleksiyRudenko — CHANGES_REQUESTED** [review]
@smillims good job.
However there is certain room for improvement.

File: submissions/smillims/oop/app.js (hunk 1 above)
Comment (smillims) [reply submissions/smillims/oop/app.js]: Yeah, it's resembles negated `block.width`

File: submissions/smillims/oop/app.js (hunk 2 above)
Comment (smillims) [reply submissions/smillims/oop/app.js]: Resolved

File: submissions/smillims/oop/app.js (hunk 3 above)
Comment (smillims) [reply submissions/smillims/oop/app.js]: Yes, it was self reference, so I removed it

File: submissions/smillims/oop/app.js (hunk 4 above)
Comment (smillims) [reply submissions/smillims/oop/app.js]: Thanks for your advice, resolved

File: submissions/smillims/oop/app.js [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+Enemy.prototype.render = function() {
+    ctx.drawImage(Resources.get(this.sprite), this.x, this.y);
+};
+
+Enemy.prototype.checkCollisions = function() {
+	if(Math.abs(this.x - player.x) < widthEnemies &&
+		Math.abs(this.y - player.y) < widthEnemies ||
+		player.y < 0){
+			setTimeout(() => player.startOver(), 120);
+	};
```
Comment (OleksiyRudenko) [inline submissions/smillims/oop/app.js:50]: Classes should not refer to any global variables directly, like global variable `player`, which is an instance of `Player` class (referring to global constants and globals provided by the gaming platform like `Resources` is OK).

https://github.com/kottans/frontend-2021-homeworks/pull/279#discussion_r546909025

Hint: Check the way `Player` gets to know about `positionPlayer.playerX`

**Review by OleksiyRudenko — CHANGES_REQUESTED** [review]
@smillims well improved.
However, one problem was solved too literraly, so that it just turned out in different location with different "participants". But the problem is all the same in its essence.

File: submissions/smillims/oop/app.js (hunk 5 above)
Comment (smillims) [reply submissions/smillims/oop/app.js]: I tried to understand how to correct it's problem and solve of this problem was refer to `player` and setup `positionPlayer.playerX`, `positionPlayer.playerX`; if it's not solve for my problem so I don't understand how to fix it if I not need to refer to `player`

Comment (OleksiyRudenko) [reply submissions/smillims/oop/app.js]: @smillims let's make a step at a time.
How does Enemy class learn about its `y` position that is defined somewhere outside the class?

Comment (smillims) [reply submissions/smillims/oop/app.js]: @OleksiyRudenko it's defined in `positionEnemyY` which defined in `allEnemies` by creating `new Enemy`

Comment (OleksiyRudenko) [reply submissions/smillims/oop/app.js]: Not exactly. The more accurate description of the situation would be:
Each Enemy's y position is stored in a global constant `positionEnemyY`, which is passed to an `Enemy` constructor as a parameter by means of `new Enemy`, and in the `Enemy` class constructor the argument for y position gets assigned to an `Enemy`'s own `y` property (which is then referred to as `this.y` in all `Enemy`'s methods).

Do you see a way to handle global `player` properly?

Comment (smillims) [reply submissions/smillims/oop/app.js]: by using `this.player.x/y`?

Comment (OleksiyRudenko) [reply submissions/smillims/oop/app.js]: Just pass global `player` to each enemy and store it in `this.player`

**Comment by smillims** [comment]
@OleksiyRudenko I think that it's finally done

File: submissions/smillims/oop/app.js [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+	width: 101,
+	height: 85
+};
+const positionEnemyY = [60, 140, 220];
+const positionPlayer = {
+	playerX: 200,
+	playerY: 400,
+};
+
+const widthEnemies = [80];
```
Comment (OleksiyRudenko) [inline submissions/smillims/oop/app.js:11]: Is there any good reason to have an array here? What will happen if we add yet another value in the list?

**Review by OleksiyRudenko — CHANGES_REQUESTED** [review]
@smillims pretty well done.
Noticed a bug: it is not safe for the player to remain on the middle bricks row even when there is no bug approaching along this row - a bug moving a level lower kills the player.

File: submissions/smillims/oop/app.js (hunk 6 above)
Comment (smillims) [reply submissions/smillims/oop/app.js]: Oh, then would be some troubles. I'm:
1) Decrease value
2) Change variable

**Review by OleksiyRudenko — APPROVED** [review]
@smillims great job!
Please test your app at every update and double check your code before submission.

**Comment by smillims** [comment]
> 
> 
> @smillims great job!
> Please test your app at every update and double check your code before submission.

Yeah, of course, big thank you for wasting your time, you are so help me in this deal. Now I be back to oop exercise

**Pull request merged**